### PR TITLE
Allow submitting additional documents to an already made proposal.

### PR DESCRIPTION
**Pull request by deiferni**
This PR allows submitting additional documents to already made proposals. It introduces the following changes:
  - add actions/views to proposal and document/documents tabs
  - track submitted documents in a separate table
  - create entries in the tracking table on proposal submission for all attached documents
  - remove entries in the tracking table when a document is deleted

### CODE REVIEW

**Comment by deiferni** [comment]
@lukasgraf plz have a look

File: opengever/meeting/command.py [hunk 1]
```diff
@@ -21,16 +26,97 @@ def execute(self):
...
+        Transporter().transport_to(
+            self.document,
+            self.submitted_document.submitted_admin_unit_id,
+            self.submitted_document.submitted_physical_path,
+            view='update-submitted-document')
+
+        submitted_version = self.document.get_current_version()
+        submitted_document = SubmittedDocument.query.get_by_source(
+            self.proposal, self.document)
+        submitted_document.submitted_version = submitted_version
```
Comment (phgross) [inline opengever/meeting/command.py]: What happens if the `transport_to` fails ?
Is there any check in the `transport_to` method I haven't realized? If not I'd could lead to a invalid or not synched document ...

File: opengever/meeting/tests/test_proposal.py [hunk 2]
```diff
@@ -181,3 +185,70 @@ def test_proposal_can_be_submitted(self, browser):
...
+        document = create(Builder('document')
+                          .within(self.dossier)
+                          .titled(u'A Document')
+                          .with_dummy_content())
+        proposal = create(Builder('proposal')
+                          .within(self.dossier)
+                          .titled(u'My Proposal')
+                          .having(committee=committee.load_model()))
+
+        self.assertFalse(proposal.is_submit_additional_documents_allowed())
```
Comment (phgross) [inline opengever/meeting/tests/test_proposal.py]: I'd move this check in to a separate test

File: opengever/meeting/tests/test_proposal.py [hunk 3]
```diff
@@ -181,3 +185,70 @@ def test_proposal_can_be_submitted(self, browser):
...
+                          .within(self.dossier)
+                          .titled(u'A Document')
+                          .with_dummy_content())
+        proposal = create(Builder('proposal')
+                          .within(self.dossier)
+                          .titled(u'My Proposal')
+                          .having(committee=committee.load_model())
+                          .relate_to(document))
+
+        proposal.execute_transition('pending-submitted')
```
Comment (phgross) [inline opengever/meeting/tests/test_proposal.py]: Using `in_state` method when creating the proposal with the builder ...

File: opengever/meeting/tests/test_proposal.py [hunk 4]
```diff
@@ -181,3 +185,70 @@ def test_proposal_can_be_submitted(self, browser):
...
+        self.assertTrue(proposal.is_submit_additional_documents_allowed())
```
Comment (phgross) [inline opengever/meeting/tests/test_proposal.py]: If you move the allowed checks in to a separate test this check is no longer necessary

File: opengever/meeting/tests/test_proposal.py [hunk 5]
```diff
@@ -181,3 +185,70 @@ def test_proposal_can_be_submitted(self, browser):
...
+        submitted_document = docs.pop()
+        self.assertEqual(0, submitted_document.get_current_version())
+
+        # create some new document versions
+        repository = api.portal.get_tool('portal_repository')
+        repository.save(document)
+        repository.save(document)
+        proposal.submit_additional_document(document)
+
+        self.assertEqual(1, submitted_document.get_current_version())
```
Comment (phgross) [inline opengever/meeting/tests/test_proposal.py]: You create two additional versions above, but the current version is only `1` ??

File: opengever/dossier/tests/test_templatedossier.py [hunk 6]
```diff
@@ -310,6 +310,7 @@ def test_enabled_actions_are_limited_in_document_tab(self):
         view = self.templatedossier.unrestrictedTraverse(DOCUMENT_TAB)
         self.assertEquals(['checkin_with_comment',
                            'checkin_without_comment',
+                           'submit_additional_documents',
```
Comment (phgross) [inline opengever/dossier/tests/test_templatedossier.py]: I think the `submit_additional_documents` action makes no sense in a template dossier.

File: opengever/meeting/command.py (hunk 1 above)
Comment (deiferni) [inline opengever/meeting/command.py]: Yes - currently we work under the assumption that only one deployment/plone-site is supported officially, but we try to prepare for multiple deployments whenever possible. So you're correct this code won't work for multiple plone-sites, but in case of one plone-site an exception is raised. (Same problem with the already merged `CopyProposalDocumentCommand`). I would suggest to  ignore this for now and implement as soon as we want to support multiple deployments. What do you think?

File: opengever/meeting/tests/test_proposal.py (hunk 3 above)
Comment (deiferni) [inline opengever/meeting/tests/test_proposal.py]: No this would not work, thats only for the plone `review_state` workflow. Note that this is our custom workflow implementation. Executing the transition calls the `Submit` transition implementation https://github.com/4teamwork/opengever.core/blob/master/opengever/meeting/proposal.py#L98 which copies the proposal's documents. I think this is also part of what this test validates.

Maybe we will implement `in_state` or `execute_transitions` for our custom builder at some point in the future but for now i'd suggest to call the transitions manually.

File: opengever/meeting/tests/test_proposal.py (hunk 5 above)
Comment (deiferni) [inline opengever/meeting/tests/test_proposal.py]: Yes thats exactly what i want to document here - the document versions run independently since these are two different document instances.

File: opengever/meeting/browser/documents/submit.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+from z3c.relationfield.schema import RelationChoice
+
+
+class SubmittableProposalFilter(CustomFilter):
+    """Custom selectable filter to only allow selection of proposals which
+    allow submission of additional documents.
+
+    """
+    def __call__(self, brain, index_data):
+        obj = brain.getObject()
```
Comment (phgross) [inline opengever/meeting/browser/documents/submit.py]: Is that unproblematic related to the performance ?

File: opengever/document/profiles/default/actions.xml [hunk 8]
```diff
@@ -76,6 +76,18 @@
       <property name="visible">True</property>
     </object>
 
+    <object name="submit_additional_documents" meta_type="CMF Action" i18n:domain="opengever.document">
+      <property name="title" i18n:translate="">Submit additional documents</property>
+      <property name="description" i18n:translate=""></property>
+      <property name="url_expr">string:submit_additional_documents:method</property>
+      <property name="icon_expr"></property>
+      <property name="available_expr">object/@@submit_additional_documents/available</property>
```
Comment (phgross) [inline opengever/document/profiles/default/actions.xml]: Does the available expression also works for other objects than the expected (for example dossiers) ?

File: opengever/meeting/command.py [hunk 9]
```diff
@@ -21,16 +26,97 @@ def execute(self):
...
+        self.document = document
+
+    def execute(self):
+        pass
+
+    def show_message(self):
+        portal = api.portal.get()
+        api.portal.show_message(
+            _(u'Document ${title} has already been submitted in that version',
+              mapping=dict(title=self.document.Title().decode('utf-8'))),
```
Comment (phgross) [inline opengever/meeting/command.py]: Using `self.document.title` instead of `self.document.Title()` would supersede the decoding ...

File: opengever/meeting/command.py [hunk 10]
```diff
@@ -21,16 +26,97 @@ def execute(self):
...
+        submitted_version = self.document.get_current_version()
+        submitted_document = SubmittedDocument.query.get_by_source(
+            self.proposal, self.document)
+        submitted_document.submitted_version = submitted_version
+
+    def show_message(self):
+        portal = api.portal.get()
+        api.portal.show_message(
+            _(u'A new submitted version of document ${title} has been created',
+              mapping=dict(title=self.document.Title().decode('utf-8'))),
```
Comment (phgross) [inline opengever/meeting/command.py]: same as above

File: opengever/meeting/tests/test_submit_additional_documents.py [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+        builder = (
+            Builder('proposal')
+            .within(self.dossier)
+            .titled(u'My Proposal')
+            .having(committee=self.committee.load_model()))
+        if attach_document:
+            builder = builder.relate_to(self.document)
+
+        proposal = create(builder)
+        proposal.execute_transition('pending-submitted')
```
Comment (phgross) [inline opengever/meeting/tests/test_submit_additional_documents.py]: Is it possible that the proposal builder supports the `in_state` method like all the other builders ?

File: opengever/meeting/browser/submitdocuments.py [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+    within its context.
+
+    Returns JSON containing the created object's path and intid.
+
+    """
+    grok.name('update-submitted-document')
+    grok.require('cmf.AddPortalContent')
+    grok.context(IDocumentSchema)
+
+    def is_submitted_document(self):
```
Comment (phgross) [inline opengever/meeting/browser/submitdocuments.py]: not used the `is_submitted_document` method is only called on the document itself.

File: opengever/document/document.py [hunk 13]
```diff
@@ -197,11 +198,24 @@ def is_checked_out(self):
...
-        return not self.is_inside_a_task()
+        return not self.is_inside_a_task() and not self.is_submitted_document()
 
     def is_inside_a_task(self):
         parent = aq_parent(aq_inner(self))
-        if ITask.providedBy(parent):
-            return True
-        else:
-            return False
+        return ITask.providedBy(parent)
```
Comment (phgross) [inline opengever/document/document.py]: Good catch :+1: :smile:

File: opengever/meeting/command.py (hunk 1 above)
Comment (phgross) [inline opengever/meeting/command.py]: Ok, but could you add a small todo comment ...

File: opengever/meeting/tests/test_proposal.py (hunk 5 above)
Comment (phgross) [inline opengever/meeting/tests/test_proposal.py]: Ah now i got it. It's the copied document not the sql representation. Everything ok !

File: opengever/meeting/tests/test_proposal.py (hunk 2 above)
Comment (deiferni) [inline opengever/meeting/tests/test_proposal.py]: :+1:

File: opengever/meeting/tests/test_proposal.py (hunk 4 above)
Comment (deiferni) [inline opengever/meeting/tests/test_proposal.py]: :+1:

File: opengever/meeting/tests/test_submit_additional_documents.py (hunk 11 above)
Comment (deiferni) [inline opengever/meeting/tests/test_submit_additional_documents.py]: yes but it would set the workflow state for the proxy object and would not call our own workflow implementation (sql-based, see above comment).

File: opengever/dossier/tests/test_templatedossier.py (hunk 6 above)
Comment (deiferni) [inline opengever/dossier/tests/test_templatedossier.py]: :+1:

File: opengever/meeting/browser/documents/submit.py (hunk 7 above)
Comment (deiferni) [inline opengever/meeting/browser/documents/submit.py]: Yes it is scoped by dossier and only gets objects only for all proposals in a dossier. This number should not exceed 100.

File: opengever/document/profiles/default/actions.xml (hunk 8 above)
Comment (deiferni) [inline opengever/document/profiles/default/actions.xml]: yes, there are two views. one for `Interface` which handles submitting generically and one for `opengever.meeting.IProposal` which is type specific.

File: opengever/meeting/command.py (hunk 9 above)
Comment (deiferni) [inline opengever/meeting/command.py]: :+1:

File: opengever/meeting/command.py (hunk 10 above)
Comment (deiferni) [inline opengever/meeting/command.py]: :+1:

File: opengever/meeting/browser/submitdocuments.py (hunk 12 above)
Comment (deiferni) [inline opengever/meeting/browser/submitdocuments.py]: :+1:

**Comment by phgross** [comment]
:shipit:

File: opengever/meeting/command.py (hunk 1 above)
Comment (deiferni) [inline opengever/meeting/command.py]: issue #728 has been updated with such a todo.
